### PR TITLE
Thea Staking Pallet

### PR DESCRIPTION
**Pull request by Gauthamastro**
It provides a way to manage thea relayers for multiple networks, it's roughly inspired by Parity's staking pallet. 

Rewards and Misbehaviour management will not be included in this PR. It will be implemented in another PR when rewards and misbehaviours are enums are defined.

Test cases are not implemented yet.

Don't review yet.

### CODE REVIEW

**Comment by felixfaisal** [comment]
This PR is ready for review @serhii-temchenko

File: pallets/thea-staking/README.md [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+\## References
+
+<!-- Original author of paragraph: @joepetrowski -->
+
+// Links to reference material, if applicable. For example, Phragmen, W3F research, etc.
+// that the implementation is based on.
+</pre></p></details>
+
+License: Unlicense
```
Comment (serhii-temchenko) [inline pallets/thea-staking/README.md]: Please remove this default readme

File: pallets/thea-staking/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,635 @@
...
+	use frame_support::{pallet_prelude::*, traits::NamedReservableCurrency};
+	use frame_system::pallet_prelude::*;
+	use sp_runtime::traits::Zero;
+
+	use crate::session::{Exposure, IndividualExposure, StakingLimits};
+
+	// Import various types used to declare pallet in scope.
+	use super::*;
+
+	//pub type StakingInfo = (Config::AccountId, Exposure<Config, <dyn Config>::AccountId>);
```
Comment (serhii-temchenko) [inline pallets/thea-staking/src/lib.rs]: Please remove

File: pallets/thea-staking/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,635 @@
...
+		type CandidateBond: Get<BalanceOf<Self>>;
+
+		/// StakingReserveIdentifier
+		#[pallet::constant]
+		type StakingReserveIdentifier: Get<<Self as pallet_balances::Config>::ReserveIdentifier>;
+
+		/// Delay to prune oldest staking data
+		type StakingDataPruneDelay: Get<SessionIndex>;
+
+		// TODO: @Faizal uncomment the code below for session change hook
```
Comment (serhii-temchenko) [inline pallets/thea-staking/src/lib.rs:96]: @felixfaisal FYI

File: pallets/thea-staking/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,635 @@
...
+	pub struct Pallet<T>(_);
+
+	// Pallet implements [`Hooks`] trait to define some logic to execute in some context.
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		// `on_finalize` is executed at the end of block after all extrinsic are dispatched.
+		fn on_finalize(_n: T::BlockNumber) {
+			// Perform necessary data/state clean up here.
+		}
+
```
Comment (serhii-temchenko) [inline pallets/thea-staking/src/lib.rs]: Please remove if no logic here

File: pallets/thea-staking/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,635 @@
...
+				// cache.
+				Weight::zero()
+			}
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight(10000)]
```
Comment (serhii-temchenko) [inline pallets/thea-staking/src/lib.rs:138]: Please add todo with issue reference about benchmarking to control this question.

File: pallets/thea-staking/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,635 @@
...
+		pub fn set_staking_limits(
+			origin: OriginFor<T>,
+			staking_limits: StakingLimits<BalanceOf<T>>,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+			<Stakinglimits<T>>::put(staking_limits);
+			Ok(())
+		}
```
Comment (serhii-temchenko) [inline pallets/thea-staking/src/lib.rs:146]: Please add documentation

File: pallets/thea-staking/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,635 @@
...
+
+		#[pallet::call_index(1)]
+		#[pallet::weight(10000)]
+		/// Adds the sender as a candidate for election and join the waitlist for selection
+		pub fn add_candidate(
+			origin: OriginFor<T>,
+			network: Network,
+			bls_key: BLSPublicKey,
```
Comment (serhii-temchenko) [inline pallets/thea-staking/src/lib.rs:159]: Please add documentation

File: pallets/thea-staking/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,635 @@
...
+				candidate,
+				stake: T::CandidateBond::get(),
+			});
+			Ok(())
+		}
+
+		#[pallet::call_index(2)]
+		#[pallet::weight(10000)]
+		/// Nominate a candidate.
+		pub fn nominate(origin: OriginFor<T>, candidate: T::AccountId) -> DispatchResult {
```
Comment (serhii-temchenko) [inline pallets/thea-staking/src/lib.rs:192]: Please add documentation

File: pallets/thea-staking/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,635 @@
...
+			let nominator = ensure_signed(origin)?;
+			Self::do_nominate(nominator, candidate)?;
+			Ok(())
+		}
+
+		#[pallet::call_index(3)]
+		#[pallet::weight(10000)]
+		/// Locks new free balance to a staker account
+		pub fn bond(origin: OriginFor<T>, amount: BalanceOf<T>) -> DispatchResult {
```
Comment (serhii-temchenko) [inline pallets/thea-staking/src/lib.rs:205]: Please add documentation

File: pallets/thea-staking/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,635 @@
...
+						&nominator,
+						amount,
+					)?;
+					<Candidates<T>>::insert(network, &candidate, exposure);
+					Self::deposit_event(Event::<T>::Bonded { candidate, nominator, amount });
+				} else {
+					return Err(Error::<T>::CandidateNotFound.into())
+				}
+			} else {
+				//FIXME: Handle this case
```
Comment (serhii-temchenko) [inline pallets/thea-staking/src/lib.rs]: Please fix

File: pallets/thea-staking/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,635 @@
...
+		pub fn bond(origin: OriginFor<T>, amount: BalanceOf<T>) -> DispatchResult {
+			let nominator = ensure_signed(origin)?;
+			Self::do_bond(nominator, amount)?;
+			Ok(())
+		}
+
+		#[pallet::call_index(4)]
+		#[pallet::weight(10000)]
+		/// Unbonds the required amount, fails if amount is greater than active stake
+		pub fn unbond(origin: OriginFor<T>, amount: BalanceOf<T>) -> DispatchResult {
```
Comment (serhii-temchenko) [inline pallets/thea-staking/src/lib.rs:218]: Please add documentation

File: pallets/thea-staking/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,635 @@
...
+			let nominator = ensure_signed(origin)?;
+			Self::do_unbond(nominator, amount)?;
+			Ok(())
+		}
+
+		#[pallet::call_index(5)]
+		#[pallet::weight(10000)]
+		/// Withdraws the unbonded funds
+		pub fn withdraw_unbonded(origin: OriginFor<T>) -> DispatchResult {
```
Comment (serhii-temchenko) [inline pallets/thea-staking/src/lib.rs:227]: Please add documentation

File: pallets/thea-staking/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,635 @@
...
+
+			Self::do_withdraw_unbonded(nominator)?;
+			Ok(())
+		}
+
+		#[pallet::call_index(6)]
+		#[pallet::weight(10000)]
+		/// Moves the candidate to Inactive state, it will also put unbonding request for all
+		/// stakers
+		pub fn remove_candidate(origin: OriginFor<T>, network: Network) -> DispatchResult {
```
Comment (serhii-temchenko) [inline pallets/thea-staking/src/lib.rs:243]: Please add documentation

File: pallets/thea-staking/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,635 @@
...
+			log::trace!(target: "runtime::thea::staking", "rotating for relayers of network {:?}", network);
+			// 1. Move queued_relayers to active_relayers
+			let active = Self::move_queued_to_active(network);
+			map.insert(network, active);
+			Self::compute_next_session(network, session_index);
+		}
+		// Increment SessionIndex
+		let new_session_index = session_index.saturating_add(1);
+		<CurrentIndex<T>>::put(new_session_index);
+		// TODO: @Faizal uncomment this when session change trait is ready.
```
Comment (serhii-temchenko) [inline pallets/thea-staking/src/lib.rs:440]: @felixfaisal FYI

**Review by serhii-temchenko — CHANGES_REQUESTED** [review]

File: pallets/thea-staking/README.md (hunk 1 above)
Comment (zktony) [reply pallets/thea-staking/README.md]: done

File: pallets/thea-staking/src/lib.rs (hunk 2 above)
Comment (zktony) [reply pallets/thea-staking/src/lib.rs]: done

File: pallets/thea-staking/src/lib.rs (hunk 4 above)
Comment (zktony) [reply pallets/thea-staking/src/lib.rs]: done

File: pallets/thea-staking/src/lib.rs (hunk 5 above)
Comment (zktony) [reply pallets/thea-staking/src/lib.rs:138]: done

File: pallets/thea-staking/src/lib.rs (hunk 6 above)
Comment (zktony) [reply pallets/thea-staking/src/lib.rs:146]: done

File: pallets/thea-staking/src/lib.rs (hunk 8 above)
Comment (zktony) [reply pallets/thea-staking/src/lib.rs:192]: done

File: pallets/thea-staking/src/lib.rs (hunk 11 above)
Comment (zktony) [reply pallets/thea-staking/src/lib.rs:218]: done

File: pallets/thea-staking/src/lib.rs (hunk 13 above)
Comment (zktony) [reply pallets/thea-staking/src/lib.rs:243]: done

File: pallets/thea-staking/src/lib.rs (hunk 12 above)
Comment (zktony) [reply pallets/thea-staking/src/lib.rs:227]: done

File: pallets/thea-staking/src/lib.rs (hunk 9 above)
Comment (zktony) [reply pallets/thea-staking/src/lib.rs:205]: done

File: pallets/thea-staking/src/lib.rs (hunk 7 above)
Comment (zktony) [reply pallets/thea-staking/src/lib.rs:159]: done

File: pallets/thea-staking/src/lib.rs (hunk 10 above)
Comment (zktony) [reply pallets/thea-staking/src/lib.rs]: done

**Review by serhii-temchenko — APPROVED** [review]

File: pallets/thea-staking/src/lib.rs (hunk 3 above)
Comment (nakul1010) [reply pallets/thea-staking/src/lib.rs:96]: Issue created: https://github.com/Polkadex-Substrate/Polkadex/issues/610

File: pallets/thea-staking/src/lib.rs (hunk 14 above)
Comment (nakul1010) [reply pallets/thea-staking/src/lib.rs:440]: issue created: https://github.com/Polkadex-Substrate/Polkadex/issues/610

**Review by nakul1010 — APPROVED** [review]
LGTM

**Pull request merged**